Add dependency injection documentation (#6978)
microsoft/fast: sites/website/src/docs/advanced/dependency-injection.md

```ts
import { DI } from "@microsoft/fast-element/di.js";

const container = DI.createContainer();
```

Now that your container is created, let's create a few dependencies to host from it.

First let's define an interface for the dependency which we'll call `MyServiceConfig`, and create a `ContextDecorator` of the same name which will take the interface as it's generic type:

```ts
import { DI } from "@microsoft/fast-element/di.js";

export interface MyServiceConfig {
    get: () => Promise<Response>;
}

export const MyServiceConfig = DI.createContext();
```

Next let's define our implementation of the service:

```ts
export class MyService implements MyServiceConfig {
    private serviceUrl: string = "http://localhost:7776";

    async get(): Promise<Response> {
        return await fetch(this.serviceUrl, { method: "GET" });
    }
}
```

> Now that we have our service defined, let's add it as a dependency to our application `App` class. Note the use of the `ContextDecorator` injected into the contructor as `@MyServiceConfig`. This will attach it to the instance with the same name.

Reviewer: spelling "contructor" = "constructor"

```ts
export class App implements AppConfig {
    constructor(
        @MyServiceConfig private readonly myService: MyServiceConfig
    ) {
        return;
    }

    public async getMyServiceStatus(): Promise<number> {
        return (await this.myService.get()).status;
    }
}
```

Finally, let's update our container by registering `MyService` with the dependency injection container. We will use the `MyServiceConfig` as the key, this will allow `App` to resolve the dependency. Note the use of `transient`, this is a utility for creating an instance each time the service is fetched. Other utilities are available, refer to the `di` API documentation for details.

```ts
import { DI, Registration } from "@microsoft/fast-element/di.js";

const { transient } = Registration;

const container = DI.createContainer();

container.register(transient(MyServiceConfig, MyService));

const responseStatus = await container.get(App).getMyServiceStatus();

console.log("Server Status:", responseStatus);
```

### Using Dependency Injection with Web Components

#### Register Dependencies Before Defining Web Components

> When using dependency injection with web components, ensure you define your container **before** you define your web components. Otherwise your web components may be intialized before the dependencies can be resolved.

Reviewer: spelling "intialized" = "initialized"

Author: Great catches, thanks! Apparently I need a spelling extension on my text editor 😭

description: FAST introduces the concept of a dependency injection container.
keywords:
    - dependency injection
    - container
    - createContainer
    - getOrSetDOMContainer
---

## Introduction

Dependency injection is a flexible & modular design pattern for passing functionality to the parts of an app that need it. If you need your project to be more maintainable and testable, this might be a good solution for you to use. The `@microsoft/fast-element` package has dependency injection utilities for regular use in JavaScript, and for injecting dependencies into web components.

### Basic Example

The first step to dependency injection is the creation of a container, this is where dependencies will be injected and resolved. There are two available ways to create a container, you can either use `createContainer()` or `getOrCreateDOMContainer()`. You can have as many containers as you like, however they will only resolve dependencies registered within their container.

```ts
import { DI } from "@microsoft/fast-element/di.js";

const container = DI.createContainer();
```

> Now that your container is created, let's create a few dependencies to inject into it.

Reviewer 2: ```suggestion
Now that your container is created, let's create a few dependencies to host from it.
```

You might consider rephrasing this slightly to clarify the container is doing the injecting, rather than being the target of dependency injection.

First let's define an interface for the dependency which we'll call `MyServiceConfig`, and create a `ContextDecorator` of the same name which will take the interface as it's generic type:

```ts
import { DI } from "@microsoft/fast-element/di.js";
import type { ContextDecorator } from "@microsoft/fast-element/context.js";

export interface MyServiceConfig {
    get: () => Promise<Response>;
}

> export const MyServiceConfig: ContextDecorator<MyServiceConfig> = DI.createContext();

Reviewer 3: I think the type annotation is unnecessary here. Passing the interface as a type parameter to createContext has the same result with less things needed to be imported.
